pkg/kubelet/kubelet.go

```diff
@@ -1938,6 +1944,15 @@ func (kl *Kubelet) syncTerminatedPod(ctx context.Context, pod *v1.Pod, podStatus
 	// generate the final status of the pod
 	// TODO: should we simply fold this into TerminatePod? that would give a single pod update
 	apiPodStatus := kl.generateAPIPodStatus(pod, podStatus)
+
+	// NOTE: resources must be unprepared BEFORE pod status is changed on the API server
+	// to avoid race conditions with the resource deallocation code in kubernetes core
+	if utilfeature.DefaultFeatureGate.Enabled(features.DynamicResourceAllocation) {
```

Comment:
If we want to ensure `UnprepareResources` is called before pod is marked as terminal in API server, then I think the options are:

1. Move call of `UnprepareResources` into `syncTerminatingPod`
2. Update the status_manager to to check that `UnprepareResources` was called and resources were released before transitioning the pod to terminal phase (https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/status/status_manager.go#L927-L933)

Reply:
In general, we want to call `UnprepareResources()`:
1. After all of a pod's containers have exited
2. Before the API server is updated to say the the pod is terminated

The current location for the call is *just* before executing:
```
	kl.statusManager.SetPodStatus(pod, apiPodStatus)
```

Which I believed to be sufficient to satisfy the constraints above.

Can you elaborate on how moving this to `syncTerminatingPod()` would satisfy these constraints better than where `UnprepareResources()` is currently being called?

Reply:
@bobbypage Thank you for the review!
After investigating this further we decided to move `UnprepareResources` call into the `syncTerminatingPod` as you've  suggested. Calling `UnprepareResources` at the earliest point (when all containers are terminated) instead of calling it just before the status update is definitely better. The change is in this PR: https://github.com/pohly/kubernetes/pull/44, please take a look.